File: backend/app/core/ratelimit.py
"""进程内滑动窗口限流器（按 IP）。

单进程内存实现，重启即清零；多 worker 部署时每个 worker 独立计数，
对自托管单 worker 场景足够。键格式 "{scope}:{ip}"。
"""

from __future__ import annotations

import time
from collections import defaultdict, deque

from app.config import settings
from app.schemas.common import AppError, ErrorCode

_attempts: dict[str, deque[float]] = defaultdict(deque)
_MAX_KEYS = 100_000


def hit(scope: str, ip: str, limit: int, window_seconds: int) -> None:
    """记录一次尝试；窗口内超过 limit 次抛 429。测试可置 RATE_LIMIT_ENABLED=false 关闭。"""
    if not settings.RATE_LIMIT_ENABLED:
        return
    key = f"{scope}:{ip}"
    now = time.monotonic()
    q = _attempts[key]
    while q and q[0] <= now - window_seconds:
        q.popleft()
    if len(q) >= limit:
        retry_after = int(q[0] + window_seconds - now) + 1
        raise AppError(
            ErrorCode.RATE_LIMITED,
            "请求过于频繁，请稍后再试",
            429,
            {"retry_after_seconds": retry_after},
        )
    q.append(now)
    if len(_attempts) > _MAX_KEYS:
        _evict(now)


def _evict(now: float) -> None:
    expired = [k for k, q in _attempts.items() if not q or q[-1] <= now - 3600]
    for k in expired:
        del _attempts[k]
    if len(_attempts) > _MAX_KEYS:
        _attempts.clear()

File: backend/internal/api/middleware/ratelimit.go
package middleware

import (
	"fmt"
	"net"
	"net/http"
	"strings"
	"sync"
	"time"

	"backend/internal/config"
	"backend/internal/model"
)

type RateLimiter struct {
	cfg      *config.Config
	mu       sync.Mutex
	attempts map[string][]float64
}

func NewRateLimiter(cfg *config.Config) *RateLimiter {
	return &RateLimiter{
		cfg:      cfg,
		attempts: make(map[string][]float64),
	}
}

func (l *RateLimiter) Hit(scope, ip string, limit, windowSeconds int) error {
	if !l.cfg.RateLimitEnabled {
		return nil
	}

	key := fmt.Sprintf("%s:%s", scope, ip)
	now := float64(time.Now().UnixNano()) / 1e9

	l.mu.Lock()
	defer l.mu.Unlock()

	history := l.attempts[key]
	threshold := now - float64(windowSeconds)

	// Filter expired timestamps
	valid := history[:0]
	for _, ts := range history {
		if ts > threshold {
			valid = append(valid, ts)
		}
	}

	if len(valid) >= limit {
		retryAfter := int(valid[0] + float64(windowSeconds) - now) + 1
		l.attempts[key] = valid
		return model.NewAppError(
			model.ErrRateLimited,
			"请求过于频繁，请稍后再试",
			429,
			map[string]any{"retry_after_seconds": retryAfter},
		)
	}

	valid = append(valid, now)
	l.attempts[key] = valid

	// Evict old keys if too many
	if len(l.attempts) > 100000 {
		for k, q := range l.attempts {
			if len(q) == 0 || q[len(q)-1] <= now-3600 {
				delete(l.attempts, k)
			}
		}
	}

	return nil
}

func GetClientIP(r *http.Request) string {
	// Check X-Forwarded-For
	if xff := r.Header.Get("X-Forwarded-For"); xff != "" {
		parts := strings.Split(xff, ",")
		if len(parts) > 0 {
			ip := strings.TrimSpace(parts[0])
			if ip != "" {
				return ip
			}
		}
	}
	// Check X-Real-IP
	if xri := r.Header.Get("X-Real-IP"); xri != "" {
		return strings.TrimSpace(xri)
	}
	host, _, err := net.SplitHostPort(r.RemoteAddr)
	if err == nil {
		return host
	}
	return r.RemoteAddr
}
